### ISSUE TITLE
Make us able to drop GNU Autotools (in favor of CMake), then drop it

### ISSUE THREAD

**Issue by hartwork** [libexpat/libexpat#330]
Before we can drop GNU Autotools, we need at least:
- [x] (1) Get MinGW support working in CMake — #334
- [x] (2) Add support for `make run-xmltest` to CMake, including test execution using Wine — depends on (1) — #334
- [x] (3) Migrate `coverage.sh` to use CMake — depends on (1) and (2), #357
- [x] (4) Migrate `qa.sh` to use CMake — depends on (1) and (2)
- [x] (5) Replace all remaining Autotools bits in Travis CI by CMake — depends on (1) and (2), potentially also on (4)
- [ ] (6) Port all remaining needed bits from `configure.ac` / `Makefile.am` to CMake
- [x] (7) Drop ~or adapt~ `expat.spec` RPM upstream package — ccd1cbc14db64f81f7404c5da37e1a9b9fa6de3b
- [x] (8) Implement equivalent to `make distcheck` for CMake — depends on (2)
- [x] (9) Port `distribute.sh` to use CMake — depends on (8)
- [ ] (10) Compare `expat_config.h` before and after —in part addressed by #437
- [ ] (11) Check CMake variables names for namespace prefix and and visibility

If you want to help with some of these, please get in touch to discuss further details before you start. Thank you!

**Comment by noloader** [comment]
After removing Autotools from Expat proper, please set up another repo `libexpat-autotools` with the sources. Folks like me will use still use Autotools, as will most of the distros. We will help maintain it.

**Comment by hartwork** [comment]
> After removing Autotools from Expat proper, please set up another repo `libexpat-autotools` with the sources. Maintain the same directory structure so we can unpack over top an expat clone. Folks like me will use still use Autotools, as will most of the distros. We will help maintain it.
> 
> Users like me don't have a choice in the matter. We need a build system that's available on a lot of systems, including modern AIX and Solaris, and older systems like CentOS 5 and Ubuntu 14. Autotools is the only build system that satisfies the requirements.
> 
> Without the broad support I will not be able to build projects like Git, GnuTLS, Hiredis, LDNS and Unbound on the platforms. Those are some of the projects that depend on Expat.

I will need to think about that.

**Comment by noloader** [comment]
>>After removing Autotools from Expat proper, please set up another repo libexpat-autotools with the sources. Maintain the same directory structure so we can unpack over top an expat clone. Folks like me will use still use Autotools, as will most of the distros. We will help maintain it...

> I will need to think about that.

Ack, thanks.

I pinged László Böszörményi (@gcsideal) on a sidebar. László is your Debian package maintainer.

This is what will happen if you remove the Autotools files and don't make them available for use. László will have to recreate the Autotools project files for use with Debian. Then, Fedoraand others who need Autotools will start using László's Autotools files.

The best situation for folks like Debian, Fedora and me is to have access to the Autotools files. Expat does not have to maintain them. All you need to do is provide a central place for folks like Debian, Fedora and me to collaborate on them.

(You *don't* want random GitHubs and BitBuckets popping up on the web trying to supply Expat project files. Only Expat should provide the files. Random GitHubs and BitBuckets offering Expat files is the Maven problem all over again).

-----

Some distros, like Gentoo, should be OK with your CMake files. I believe they have a workflow that includes CMake.

And you may not realize it, but if Expat was CMake-only it would currently be broken on AIX and MacPort. CMake cannot build on AIX or MacPorts at the moment. It would also be broken on Ubuntu 14 because Ubuntu 14 does not meet the minimum CMake requirements.

**Comment by DerDakon** [comment]
I wonder how you are able to build e.g. KDE software which uses CMake exclusively.

There is at least one AIX machine on the CMake dashboard that is able to build CMake and usually passes the tests. What issue are you seeing? Please point me to the CMake tracker, this does not belong here.

The MacPort issue looks like there _may_ be a screwed toolchain update. I have relayed this to the [CMake tracker](https://gitlab.kitware.com/cmake/cmake/issues/20462) as noone seems to have cared to do it before.

**Comment by hartwork** [comment]
> Users like me don't have a choice in the matter. We need a build system that's available on a lot of systems, including modern AIX and Solaris, and older systems like CentOS 5 and Ubuntu 14. Autotools is the only build system that satisfies the requirements.

@noloader could you elaborate why these environments have no chance of getting recent CMake packaged and working?

> This is what will happen if you remove the Autotools files and don't make them available for use. László will have to recreate the Autotools project files for use with Debian. Then, Fedora and others who need Autotools will start using László's Autotools files.
> 
> The best situation for folks like Debian, Fedora and me is to have access to the Autotools files. 

You're effectively saying that Debian and Fedora packages cannot package CMake-based software without replacing the build system with something else. What is your source for this?
I would like to provide official sources saying contrary:
- Federa: https://docs.fedoraproject.org/en-US/packaging-guidelines/CMake/
- Debian: https://www.debian.org/doc/manuals/debmake-doc/ch05.en.html#cmake

**Comment by noloader** [comment]
> @noloader could you elaborate why these environments have no chance of getting recent CMake packaged and working?

I don't build compilers. I use the compiler and glibc that came with the platform.


> You're effectively saying that Debian and Fedora packages cannot package CMake-based software without replacing the build system with something else. What is your source for this?

Experience based on other projects I help maintain.

**Comment by gcsideal** [comment]
> You're effectively saying that Debian and Fedora packages cannot package CMake-based software without replacing the build system with something else. What is your source for this?
Just checked and it seems our CMake packaged version is quite recent. Will discuss details with @noloader .

**Comment by hartwork** [comment]
> > @noloader could you elaborate why these environments have no chance of getting recent CMake packaged and working?
> 
> I don't build compilers. I use the compiler and glibc that came with the platform.

Could you elaborate why you want to combine too-old compilers with modern software releases? I would get if (a) you'd backport security fixes to environments from the past or (b) use or backport recent compilers to run modern software. Why this balancing act?

**Comment by diizzyy** [comment]
The downside is that cmake depends on (lib)expat which isn't ideal, that's probably why many still are on autotools. Any ideas about that?

**Comment by hartwork** [comment]
Hi @diizzyy , I'm aware of that loop (and it's one reason why I haven't moved forward with dropping Autotools, yet). I don't expect much of a problem with *binary*-based distributions. With regard to *source*-based distributions, the loop is not impossible to break but not ideal, I agree. The [related thread on the `gentoo-dev` mailing list](https://marc.info/?l=gentoo-dev&m=157670297302319) may be of interest. What would you do if you were in Expat's shoes?

**Comment by gcsideal** [comment]
> I'm aware of that loop (and it's one reason why I haven't moved forward with dropping Autotools, yet). [...] What would you do if you were in Expat's shoes?

There's the expat -> docbook2x -> libxml-sax-expat-perl -> libxml-parser-perl -> libexpat1 loop as well (in Debian). A stripped down version of expat would be nice without too much build dependency (built with Autotools). With that dependent packages could be built and then the full featured expat would be built with whatever build system would be best in the modern world.

**Comment by hartwork** [comment]
> There's the expat -> docbook2x -> libxml-sax-expat-perl -> libxml-parser-perl -> libexpat1 loop as well (in Debian).

Hi! The release tarballs come with a pre-built `xmlwf.1` man page by now. I stopped rebuilding the man page in Gentoo packaging, because the man page is not code. Is that an option to you as well?

> A stripped down version of expat would be nice without too much build dependency (built with Autotools). With that dependent packages could be built and then the full featured expat would be built with whatever build system would be best in the modern world.

Let me think about that.

**Comment by DerDakon** [comment]
CMake can be bootstrapped without external dependencies, it bundles everything it needs.

**Comment by hartwork** [comment]
> CMake can be bootstrapped without external dependencies, it bundles everything it needs.

There are security concerns with that approach, because for every fixed vulnerability in Expat, distros now need to update Expat and CMake. See https://wiki.gentoo.org/wiki/Why_not_bundle_dependencies .

**Comment by DerDakon** [comment]
I said bootstrapped. Once that is done you can build expat, build the real CMake and drop the other one.

**Comment by hartwork** [comment]
Point taken about "bootstrapped". I was aware of that option with two CMake packages (as you know from our discussion on `gentoo-dev`).

**Comment by hartwork** [comment]
> A stripped down version of expat would be nice without too much build dependency (built with Autotools). With that dependent packages could be built and then the full featured expat would be built with whatever build system would be best in the modern world.

@gcsideal wouldn't that effectively mean keeping Autotools, i.e. no progress and continue maintaining two build systems?  What did you picture as the minimal build system that still satisfies all distro packaging needs?  I'm not sure I get the idea then, please elaborate.

PS: I have added milestone 2.3.1 to this ticket now, e.g. to signal that a release with billion laughs attack protection is planned to come prior to potentially dropping the GNU Autotools build system.

**Comment by 1480c1** [comment]
> wouldn't that effectively mean keeping Autotools, i.e. no progress and continue maintaining two build systems? What did you picture as the minimal build system that still satisfies all distro packaging needs? I'm not sure I get the idea then, please elaborate.

bit old but, I think the idea is to make a separate repo under libexpat that contains only the autotool related files that they could just copy and layer over a local libexpat repo and isn't necessarily maintained by the members but rather co-maintained by distro maintainers who might still need the autotool files just so that way they sill have a "canonical" autotools repo they can centrally co-maintain instead of keeping a bunch of local patches in each distro's source repo that all have the same basic functionality.

**Comment by hartwork** [comment]
@1480c1 for the record I'm not a big fan of that idea.

**Comment by 1480c1** [comment]
For the record, me too

**Comment by hartwork** [comment]
Situation "cyclic dependency between libexpat and CMake" remains unchanged and unsolved. As a result, the GNU Autotools build system will not go away from libexpat in the near future. I am therefore closing this issue as "won't fix any time soon".

**Comment by gruenich** [comment]
I reached out to the maintainer of CMake.
https://discourse.cmake.org/t/impact-of-libexpat-dropping-automake/11411

He confirmed that CMake can be built without libexpat.
1. CMake can bootstrap and compile libexpat from CMake's sources.
2. This bootstrapped CMake can be used to build libexpat.
3. A usual CMake can be built with libexpat as a dependency.

**Comment by diizzyy** [comment]
@gruenich 
Which is an issue as many package repos etc do not approve of bundled libraries due to security and maintenance

**Comment by hartwork** [comment]
> @gruenich Which is an issue as many package repos etc do not approve of bundled libraries due to security and maintenance

Full ack — https://wiki.gentoo.org/wiki/Why_not_bundle_dependencies is related.

The [key quote](https://discourse.cmake.org/t/impact-of-libexpat-dropping-automake/11411/3) to me is:

> There is currently no option to build CMake without the features that rely on expat. However, CMake can be built without using a system-provided expat because it vendors a copy of the source.

So it would force distros into quite a mess of a situation.

**Comment by gruenich** [comment]
> Which is an issue as many package repos etc do not approve of bundled libraries due to security and maintenance

That is why I mentioned point 3., building CMake without bundled dependencies.

It is a matter of weighting. Are distributions responsible and should provide mechanism to bootstrap potential dependency cycles? I hope they can already cope with that for compilers and other build systems. Alternately, will libexpat carry the burden of maintaining two build systems just for this edge case? The CMake part of libexpat is unnecessary bloated and slow to mimic Autoconf's compatibility of last millennium's C compilers. Autotools are slow, complex, and can carry security risks as the xz backdoor demonstrated. They lack man power and have difficulties to stay compatible with newer compiler releases.

**Comment by noloader** [comment]
> Are distributions responsible and should provide mechanism to bootstrap potential dependency cycles? 

I don't believe Debian or Fedora will include it if it is precompiled. And there are some modules that fail to build from sources (FTBFS). The module's authors tell the distros "use our precompiled version", and that gets rejected by the distros. 

An example of a rejection is gecko-driver. Gecko-driver is Firefox's Selenium WebDriver. It takes some sort of special pixie dust to build gecko-driver. Debian and Fedora will not include gecko-driver with Firefox. From <https://bugzilla.redhat.com/show_bug.cgi?id=1771769#c4>:

> ... The most important thing is that the current spec just ships a pre-built binary. The package as-is MUST NOT be approved! I think Thomas should focus on that first.
> 
> Also it might be non-trivial to package geckodriver which requires a few rust packages. But anyway: Build from source first before tackling cosmetic issues.

**Comment by gruenich** [comment]
Nobody suggests precompiled binaries. It is a two-step approach. Build CMake from source with the bundled libraries (from source). Use this to build libexpat. Now compile CMake using its external dependencies.

**Comment by hartwork** [comment]
> Autotools [..] carry security risks as the xz backdoor demonstrated.

@gruenich it could be worth noting that any build environment that already has recent-enough Autotools (and its dependencies) installed can take a [tarball snapshot of the `R_2_6_2` Git tag](https://github.com/libexpat/libexpat/archive/refs/tags/R_2_6_2.tar.gz) and run `./buildconf.sh` to generate the needed Autotools files rather than using the `make dist` release archives with these files pre-generated.

> Build CMake from source with the bundled libraries (from source). Use this to build libexpat. Now compile CMake using its external dependencies.

@gruenich I have no current plans of forcing 50+ distros to start maintaining two copies of the [CMake package](https://gitweb.gentoo.org/repo/gentoo.git/tree/dev-build/cmake) just to be able to build libexpat from source the two-step way you described.
